lib/shared/src/chat/preamble.ts

```diff
@@ -1,14 +1,28 @@
-import { type PromptString, ps } from '../prompt/prompt-string'
+import { type PromptString, ps, psDedent } from '../prompt/prompt-string'
 import type { Message } from '../sourcegraph-api'
 
+const DEFAULT_PREAMBLE = ps`You are Cody, an AI coding assistant from Sourcegraph.`
+
+/**
+ * For chat, we add an additional preamble to encourage the model to
+ * produce code blocks that we can associate with existing file paths.
+ * We want to read these file paths to support applying code directly to files from chat.
+ */
+const CHAT_PREAMBLE = psDedent`
```

Comment:
```suggestion
const SMART_APPLY_PREAMBLE = ps`If your answer contains fenced code blocks in Markdown, include the full file path in the code block tag using this structure: \`\`\`$LANGUAGE:$FILEPATH\n\`\`\`.`
const CHAT_PREAMBLE = DEFAULT_PREAMBLE.join(SMART_APPLY_PREAMBLE)
```

I think we don't need to say that this is an additional rules in system prompt. We can probably update it to something like this? ( Haven't tested it though!)

Reply:
Happy to try anything, will run an eval on this change and we can see!

Reply:
@abeatrix Applied this change and made a couple of more tweaks, looks pretty good!

Eval is here: https://github.com/sourcegraph/cody-leaderboard/pull/16

My comments here are still relevant: https://github.com/sourcegraph/cody-leaderboard/pull/16#issuecomment-2304581865

I think we can go ahead and merge this, we can always tweak later but let's get a patch out today as we may be both OOO next week